keps/sig-network/20200204-cni-verification-rearchitecture.md

```diff
@@ -0,0 +1,622 @@
+ 
+tests.  These tests use almost identical harnesses, with a subtle `},{` clause
+differentiating the stacked peers in network policy (an *or* selector) vs. a combined peer policy.
+ 
+```
+ 
+                    Ingress: []networkingv1.NetworkPolicyIngressRule{{
+                        From: []networkingv1.NetworkPolicyPeer{
+                            {
+                                // TODO add these composably, so that can be disambiguated from combo networkpolicypeer
+                                PodSelector: &metav1.LabelSelector{
+                                    MatchLabels: map[string]string{
+                                        "pod-name": "client-b",
+                                    },
+                                },
+                            },
+                            {
+                                NamespaceSelector: &metav1.LabelSelector{
+                                    MatchLabels: map[string]string{
+                                        "ns-name": nsBName,
+                                    },
+                                },
+                            },
+                        },
+                    }},
+```
+ 
+The AND test is obviously more selective, although it is tricky to tell from the struct that
+it has been correctly written to be different from the OR test.
+ 
+```
+                    Ingress: []networkingv1.NetworkPolicyIngressRule{{
+                        From: []networkingv1.NetworkPolicyPeer{
+                            {
+                                PodSelector: &metav1.LabelSelector{
+                                    MatchLabels: map[string]string{
+                                        "pod-name": "client-b",
+                                    },
+                                },
+                                // because we lack {,} , this is a single Peer selecting
+                                // pods, from namespaces selected by namespace selector.
+                                // This is difficult to verify for correctness at a
+                                // glance, due to the verbosity of the struct.
+                                NamespaceSelector: &metav1.LabelSelector{
+                                    MatchLabels: map[string]string{
+                                        "ns-name": nsBName,
+                                    },
+                                },
+                            },
+                        },
+```
+
+In contrast, we can express the same and test using the API (including its entire connectivity matrix) in this proposal as follows:
+
+```
+	builder := &NetworkPolicySpecBuilder{}
+	builder = builder.SetName("myns", "allow-podb-in-nsb").SetPodSelector(map[string]string{"pod": "b"})
+	builder.SetTypeIngress()
+	builder.AddIngress(nil, &p80, nil, nil, map[string]string{"pod-name": "b"}, nil, map[string]string{"ns-name": "b"}, nil)
```

Comment:
... it's really not clear to me that creating a new syntax, used only in the networkpolicy tests, is really going to make things more understandable.

Reply:
sure, could go either way assuming your still ok with this approach ?

Reply:
I think the test cases should use either `networkingv1.NetworkPolicy` structs or inline YAML, because that's the real/official representation of NetworkPolicies.

If your primary concern is that people reading the test cases might misunderstand the policies then this could be addressed via comments, or *maybe* by building the policies in pieces, eg:

	policy := &networkingv1.NetworkPolicy{
		ObjectMeta: metav1.ObjectMeta{
			Name: "allow-ns-b-via-namespace-selector-or-client-b-via-pod-selector",
		},
		Spec: networkingv1.NetworkPolicySpec{
			PodSelector: metav1.LabelSelector{
				MatchLabels: map[string]string{
					"pod-name": podServerLabelSelector,
				},
			},
			Ingress: []networkingv1.NetworkPolicyIngressRule{
				From: []networkingv1.NetworkPolicyPeer{},
			},
		},
	}
	policy.Spec.Ingress.From = append(policy.Spec.Ingress.From,
		networkingv1.NetworkPolicyPeer{
			PodSelector: &metav1.LabelSelector{
				MatchLabels: map[string]string{
					"pod-name": "client-b",
				},
			},
		},
	}
	policy.Spec.Ingress.From = append(policy.Spec.Ingress.From,
		networkingv1.NetworkPolicyPeer{
			NamespaceSelector: &metav1.LabelSelector{
				MatchLabels: map[string]string{
					"ns-name": nsBName,
				},
			},
		},
	}

actually, probably more comments are good either way. But anyway, if the problem is that NetworkPolicies are too hard to understand, then that's best addressed by making NetworkPolicies easier to understand (ie, v2), not by avoiding using NetworkPolicy directly in the tests.